onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -793,19 +797,16 @@ of [N, 0] then [N, 0].
           "An optional string. Token's regular expression in basic POSIX format"
           " (http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap09.html#tag_09_03)."
           " If set, tokenizer may produce tokens matching the specified pattern. Note that one and only of"
-          " 'tokenexp' and 'separators' should be set.",
+          " 'tokenexp' and 'sepexp' should be set.",
           AttributeProto::STRING,
           OPTIONAL)
       .Attr(
           "separators",
-          "an optional list of strings (type: AttributeProto::STRINGS), each single string in this attribute is a separator."
+          "an optional list of strings attribute that contains a separator - regular expressions to match separators"
           " Two consecutive segments in X connected by a separator would be divided into two tokens."
           " For example, if the input is \"Hello World!\" and this attribute contains only one space character,"
           " the corresponding output would be [\"Hello\", \"World!\"]. To achieve character-level tokenization,"
-          " one should set the separators to [\"\"], which contains only one empty string."
-          " If 'separators' is a L-element array, there will be L rounds of tokenization using one stop word."
-          " More specifically, in the first round, the first element in 'separators' is used to tokenize each string in the input."
-          " Then, the second element in 'separators' will be used to tokenize the resulted strings produced at the first round.",
+          " one should set the 'separators' to [\"\"], which contains an empty string.",
```

Comment:
I am not sure if we can do this. Protobuf's default string is `""`, so we will do char-level tokenization by default. Maybe using a more complicated pattern such as `token_expression="."` to denote char-tokenization?

Reply:
I vote for parameter with explicit meaning. A parameter telling which strategy to follow (char, word with separator, word with regular expression), a string specifying either a list of separators or a regular expression.

Reply:
Protobuf has nothing to do with it and it is not true. Protobuf may not have a member at all and that's why it has has_* methods.
The array of strings will never have empty strings populated as default. It can't. Especially, if the attribute not set in its entirety.

---
In reply to: [276864749](https://github.com/Microsoft/onnxruntime/pull/857#discussion_r276864749) [](ancestors = 276864749)

Reply:
We chose to preserve the existing parameters with backwards compatibility both in names and behavior.

---
In reply to: [276956862](https://github.com/Microsoft/onnxruntime/pull/857#discussion_r276956862) [](ancestors = 276956862)